unison-core/src/Unison/ABT.hs

```diff
@@ -690,47 +699,55 @@ hashComponents termFromHash termsByName = let
-    Embed fa -> Hashable.hash1 hashCycle hash fa
-
 -- | We ignore annotations in the `Term`, as these should never affect the
 -- meaning of the term.
 hash :: forall f v a h . (Functor f, Hashable1 f, Eq v, Show v, Ord h, Accumulate h)
      => Term f v a -> h
 hash = hash' [] where
-  hash' :: [Either [v] v] -> Term f v a -> h
-  hash' env (Term _ _ t) = case t of
-    Var v -> maybe die hashInt ind
-      where lookup (Left cycle) = v `elem` cycle
-            lookup (Right v') = v == v'
-            ind = findIndex lookup env
-            hashInt :: Int -> h
-            hashInt i = Hashable.accumulate [Hashable.Nat $ fromIntegral i]
-            die = error $ "unknown var in environment: " ++ show v
-                        ++ " environment = " ++ show env
-    Cycle (AbsN' vs t) -> hash' (Left vs : env) t
-    -- Cycle t -> hash' env t
-    Abs v t -> hash' (Right v : env) t
-    Tm t -> Hashable.hash1 (hashCycle env) (hash' env) t
-
-  hashCycle :: [Either [v] v] -> [Term f v a] -> ([h], Term f v a -> h)
-  hashCycle env@(Left cycle : envTl) ts | length cycle == length ts =
-    let
-      permute p xs = case Vector.fromList xs of xs -> map (xs !) p
-      hashed = map (\(i,t) -> ((i,t), hash' env t)) (zip [0..] ts)
-      pt = fst <$> sortOn snd hashed
-      (p,ts') = unzip pt
-    in case map Right (permute p cycle) ++ envTl of
-      env -> (map (hash' env) ts', hash' env)
-  hashCycle env ts = (map (hash' env) ts, hash' env)
+
+hash' :: forall f v a h . (Functor f, Hashable1 f, Eq v, Show v, Ord h, Accumulate h)
+      => [Either [v] v] -> Term f v a -> h
+hash' env (Term _ _ t) = case t of
+  Var v -> maybe die hashInt ind
+    where lookup (Left cycle) = v `elem` cycle
+          lookup (Right v') = v == v'
+          ind = findIndex lookup env
+          hashInt :: Int -> h
+          hashInt i = Hashable.accumulate [Hashable.Nat $ fromIntegral i]
+          die = error $ "unknown var in environment: " ++ show v
+                      ++ " environment = " ++ show env
+  Cycle (AbsN' vs t) -> hash' (Left vs : env) t
+  -- Cycle t -> hash' env t
+  Abs v t -> hash' (Right v : env) t
+  Tm t -> Hashable.hash1 (hashCycle env) (hash' env) t
+  where
+    hashCycle :: [Either [v] v] -> [Term f v a] -> ([h], Term f v a -> h)
+    hashCycle (Left cycle : env) ts | length cycle == length ts =
+      let (ts', env') = doHashCycle env (zip cycle ts)
+      in (ts', hash' env')
+    hashCycle env ts = (map (hash' env) ts, hash' env)
+
+-- | @doHashCycle env terms@ hashes cycle @terms@ in environment @env@, and returns the canonical ordering of the hashes
+-- of those terms, as well as an updated environment with each of the terms' bindings in the canonical ordering.
+doHashCycle ::
+  forall a f h v.
+  (Accumulate h, Eq v, Hashable1 f, Ord h, Show v) =>
+  [Either [v] v] ->
+  [(v, Term f v a)] ->
+  ([h], [Either [v] v])
+doHashCycle env namedTerms =
+  (map (hash' newEnv) permutedTerms, newEnv)
```

Comment:
Writing down some possible further optimizations before I forget:

It seems like we're still hashing a given term 2 complete times, once here where each Variable in the component is distinct in the environment, and once on line 746 where all variables in the component are considered equivalent; that got me thinking: The hash from 746 should still serve as an adequate content-based hash for the _structure_ of the term, just with the exception that different variables within the same cycle aren't yet distinguished. So when we re-hash here, all we really need is to introduce uniqueness with regard to that.

I'm wondering if it might speed things up if the initial hash on 746 also returns the list of all the uses of self-referential variables in the order they appear, then for the hash on 739 we do something like this:

```
accumulate (Hashed existingHash : tokens (Int . findIndex permutedNames <$> usedVars))
```

Super rough idea, and it would change the hashes as well as make this code trickier to read, so I'm really not sure it'd even be worth it.

We could also investigate annotate subtrees with their hashes and re-using any that are still valid in the second hash, since subtrees which don't have self-references won't change from the first hash to the second, but that's another rather significant change.